docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+s3gw image is not cached, the system should pay the time needed to download
+the image from the registry before starting it.
+
+## Investigation's Rationale
+
+The 3 models described above have different performances and different implementation
+efforts. While the *Active-Active* model is expected to require a significant
+development effort due to its inherent complex nature, the *Passive-Active* model
+seems to offer a lower degree of implementation difficulty.
+
+In particular, the *One Active pipeline* model, is expected to work with nothing
+but the primitives normally available on any Kubernetes cluster.
+
+Given that the backend s3gw architecture is composed by:
+
+- one *stateless* `radosgw` process associated with a *stateful* `PV`
+
+it is supposed that, on `radosgw` process failure, a simple reschedule
+of its POD is enough to fulfill this simple HA model.
+A clarification is needed here: the `PV` *state* is guaranteed by a third party
+service, that is: **Longhorn**.
+
+Longhorn ensures that a `PV` (along with its content) is always available on
+any cluster's node, so that, a POD can mount it regardless of its allocation on the
+cluster.
+
+Relying on this assumption, while we are still in the process to decide what model
+we want to actually engage, the *One Active pipeline* model seems for sure
+the easiest thing to explore and to investigate.
+
+## Failure cases
+
+The `PV` state is kept coherent by Longhorn, so errors at this level are assumed
+NOT possible (application level corruptions to the `PV`'s state ARE possible).
+
+What are the failure cases that can happen for s3gw?
+
+Let's examine the following scenarios:
+
+1. `radosgw`'s POD failure or rescheduling
+2. Cluster's node failure
+3. `radosgw`'s failure due to a bug not related to a certain input pattern
+4. `radosgw`'s failure due to a bug related to a certain input pattern
+5. `PV` Data corruption at application level due to `radosgw`'s bug
+6. `PV` Data corruption at application level due to `radosgw`'s anomalous exit
+
+We are supposing all these bugs or conditions to be fatal for the s3gw's process
+so that they trigger an anomalous exit.
+
+### radosgw's POD failure
+
+This case is when the `radosgw` process stops due to a failure of its POD.
+This case applies also when Kubernetes decides to reschedule the POD
+to another node, eg: when the node goes low on resources.
+
+This can be thought as an infrastructure issue independent to the s3gw.
+In this case, the *One Active pipeline* model fully restores the service by
+rescheduling a new POD somewhere in the cluster.
+
+### Cluster's node failure
+
+This case is when the `radosgw` process stops due to a cluster's node failure.
+The *One Active pipeline* model fully restores the service by
+rescheduling a new POD somewhere in the cluster.
+
+### radosgw's failure due to a bug not related to a certain input pattern
+
+This case is when the `radosgw` process crashes due to a bug not directly
```

Comment:
I think that make these cases explicit is useful for a theoretical reasoning on what scenarios we can actually think to solve with an HA model.
This case is clearly outside what an HA model can handle; we must expect for this case that the `back off` mechanism is the only mitigation for a possible restart loop should occur.

Reply:
Fair, we can rely on the back-off loop in some cases (though _maybe_ the admin wants to disable that so that, once the possibly faulty-itself client stops, s3gw restarts as quickly as possible).

My initial suggestion would be to focus on node crashes (or their equivalent, such as OOM etc), and restarts due to that, since that's also what clients are most likely to test.

Reply:
> though maybe the admin wants to disable that so that, once the possibly faulty-itself client stops, s3gw restarts as quickly as possible
afaik, the backoff timer can not be disabled (or tuned) without modifying K8s code.